File: C1.UWP.FlexGrid/VB/FlexGridSamples/CellFactory/RatingCell.vb
Imports Windows.UI.Xaml.Shapes
Imports Windows.UI.Xaml.Media.Animation
Imports Windows.UI.Xaml.Media
Imports Windows.UI.Xaml.Data
Imports Windows.UI.Xaml.Controls
Imports Windows.UI.Xaml
Imports Windows.UI
Imports System

''' <summary>
''' Cell that shows a rating value as an image with stars.
''' </summary>
Public Class RatingCell
    Inherits StackPanel
    Const MAXRATING As Integer = 5
    Const OFF As Double = 0.2
    Const [ON] As Double = 1.0

    ''' <summary>
    ''' Identifies the <see cref="ItemsSource"/> dependency property.
    ''' </summary>
    Public Shared ReadOnly RatingProperty As DependencyProperty = DependencyProperty.Register("Rating", GetType(Integer), GetType(RatingCell), New PropertyMetadata(0, AddressOf OnRatingChanged))

    Public Sub New()
        Orientation = Orientation.Horizontal
        Dim i As Integer = 0
        While i < 5
            Dim star As New ContentControl()
            star.Width = 20
            star.Height = 20
            star.ContentTemplate = GetStarPath()
            star.Opacity = OFF
            AddHandler star.PointerPressed, AddressOf img_PointerPressed
            Children.Add(star)
            i += 1
        End While
    End Sub

    Public Property Rating() As Integer
        Get
            Return CType(GetValue(RatingProperty), Integer)
        End Get
        Set
            SetValue(RatingProperty, Value)
        End Set
    End Property
    Sub img_PointerPressed(sender As Object, e As Windows.UI.Xaml.Input.PointerRoutedEventArgs)
        ' calculate rating based on the index of the star
        Dim star As ContentControl = TryCast(sender, ContentControl)
        Dim cell As RatingCell = TryCast(star.Parent, RatingCell)
        Dim index As Integer = cell.Children.IndexOf(star)
        If index > 0 OrElse e.GetCurrentPoint(star).Position.X > star.Width / 3 Then
            index += 1
        End If

        ' apply the new rating
        cell.Rating = index
        Animate(star)
    End Sub

    Shared Function GetStarPath() As DataTemplate
        Return TryCast(TemplateCell.CustomTemplatesDictionary("StarIcon"), DataTemplate)
    End Function

    Shared Sub OnRatingChanged(d As DependencyObject, e As DependencyPropertyChangedEventArgs)
        Dim cell As RatingCell = CType(d, RatingCell)
        Dim i As Integer = 0
        While i < cell.Children.Count
            cell.Children(i).Opacity = If(i < cell.Rating, [ON], OFF)
            i += 1
        End While
    End Sub

    Shared _animate As FrameworkElement
    Shared _sb As Storyboard
    Sub Animate(star As FrameworkElement)
        ' create storyboard
        If _sb Is Nothing Then
            _sb = New Storyboard()
            _sb.Duration = New Duration(TimeSpan.FromMilliseconds(20))
            AddHandler _sb.Completed, AddressOf sb_Completed
        End If

        ' suspend current animation if any
        If _animate IsNot Nothing Then
            _sb.[Stop]()
            _animate.RenderTransform = Nothing
        End If

        ' start new animation
        star.RenderTransform = New ScaleTransform() With {
                .ScaleX = 2,
                .ScaleY = 2,
                .CenterX = star.Width / 2,
                .CenterY = star.Height / 2
            }
        _animate = star
        _sb.Begin()
    End Sub

    Sub sb_Completed(sender As Object, e As Object)
        Dim st As ScaleTransform = TryCast(_animate.RenderTransform, ScaleTransform)
        If st IsNot Nothing Then
            If st.ScaleX > 1 Then
                st.ScaleX -= 0.1
                st.ScaleY -= 0.1
                _sb.Begin()
            End If
            Return
        End If
    End Sub
End Class

File: C1.UWP.FlexGrid/CS/FlexGrid101/CellFactory/RatingCell.cs
using System;
using Windows.UI;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Media.Animation;
using Windows.UI.Xaml.Shapes;

namespace FlexGrid101
{
    /// <summary>
    /// Cell that shows a rating value as an image with stars.
    /// </summary>
    public class RatingCell : StackPanel
    {
        const int MAXRATING = 5;
        const double OFF = 0.2;
        const double ON = 1.0;

        /// <summary>
        /// Identifies the <see cref="ItemsSource"/> dependency property.
        /// </summary>
        public static readonly DependencyProperty RatingProperty =
            DependencyProperty.Register(
                "Rating",
                typeof(int),
                typeof(RatingCell),
                new PropertyMetadata(0, OnRatingChanged));

        public RatingCell()
        {
            Orientation = Orientation.Horizontal;
            for (int i = 0; i < 5; i++)
            {
                var star = new ContentControl();
                star.Width = 20;
                star.Height = 20;
                star.ContentTemplate = GetStarPath();
                star.Opacity = OFF;
                star.PointerPressed += img_PointerPressed;
                Children.Add(star);
            }
        }

        public int Rating
        {
            get { return (int)GetValue(RatingProperty); }
            set { SetValue(RatingProperty, value); }
        }
        void img_PointerPressed(object sender, Windows.UI.Xaml.Input.PointerRoutedEventArgs e)
        {
            // calculate rating based on the index of the star
            var star = sender as ContentControl;
            RatingCell cell = star.Parent as RatingCell;
            int index = cell.Children.IndexOf(star);
            if (index > 0 || e.GetCurrentPoint(star).Position.X > star.Width / 3)
            {
                index++;
            }

            // apply the new rating
            cell.Rating = index;
            Animate(star);
        }

        static DataTemplate GetStarPath()
        {
            return TemplateCell.CustomTemplatesDictionary["StarIcon"] as DataTemplate;
        }

        static void OnRatingChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
            RatingCell cell = (RatingCell)d;
            for (int i = 0; i < cell.Children.Count; i++)
            {
                cell.Children[i].Opacity = i < cell.Rating ? ON : OFF;
            }
        }

        static FrameworkElement _animate;
        static Storyboard _sb;
        void Animate(FrameworkElement star)
        {
            // create storyboard
            if (_sb == null)
            {
                _sb = new Storyboard();
                _sb.Duration = new Duration(TimeSpan.FromMilliseconds(20));
                _sb.Completed += sb_Completed;
            }

            // suspend current animation if any
            if (_animate != null)
            {
                _sb.Stop();
                _animate.RenderTransform = null;
            }

            // start new animation
            star.RenderTransform = new ScaleTransform() { ScaleX = 2, ScaleY = 2, CenterX = star.Width / 2, CenterY = star.Height / 2 };
            _animate = star;
            _sb.Begin();
        }

        void sb_Completed(object sender, object e)
        {
            var st = _animate.RenderTransform as ScaleTransform;
            if (st != null)
            {
                if (st.ScaleX > 1)
                {
                    st.ScaleX -= .1;
                    st.ScaleY -= .1;
                    _sb.Begin();
                }
                return;
            }
        }
    }
}
